### PR TITLE
JDK-8296764 NMT: reduce loads in os::malloc

### PR DESCRIPTION
**Pull request by tstuefe**
We read the NMT level at every ps::malloc/realloc/free etc, even if NMT is off. That is unavoidable. But we also read a second variable that shadows the NMT level `_nmt_was_initialized`. That state is synonymous with NMT level "unknown," so there is no need for this second load.

Furthermore, NMT level is volatile. That had been necessary in older times since we used to shut down NMT on resource exhaustion. But since [JDK-8256844](https://bugs.openjdk.org/browse/JDK-8256844), NMT level is fixed at initialization, so we can remove the volatile specifier.

Patch 
- removes `_nmt_was_initialized` and redirects it to `MemTracker::is_initialized()`, which gets inlined to a load of NMT level. 
- makes NMT level non-volatile
- removes _is_nmt_env_valid, which had been dead code

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296764](https://bugs.openjdk.org/browse/JDK-8296764): NMT: reduce loads in os::malloc

### Reviewers
 * [Andrew Dinn](https://openjdk.org/census#adinn) (@adinn - **Reviewer**) ⚠️ Review applies to [9d5cac2c](https://git.openjdk.org/jdk/pull/11080/files/9d5cac2c0659413c54fa7f36aa3f8b382c3bc1fb)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11080/head:pull/11080` \
`$ git checkout pull/11080`

Update a local copy of the PR: \
`$ git checkout pull/11080` \
`$ git pull https://git.openjdk.org/jdk pull/11080/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11080`

View PR using the GUI difftool: \
`$ git pr show -t 11080`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11080.diff">https://git.openjdk.org/jdk/pull/11080.diff</a>

</details>

### CODE REVIEW

**Comment by tstuefe** [comment]
x86 test error in compiler/c2/TestVerifyGraphEdges.java unrelated

File: src/hotspot/share/services/nmtPreInit.hpp [hunk 1]
```diff
@@ -30,6 +30,7 @@
 #ifdef ASSERT
 #include "runtime/atomic.hpp"
 #endif
+#include "services/memTracker.hpp" // for MemTracker::is_initialized()
```
Comment (adinn) [inline src/hotspot/share/services/nmtPreInit.hpp:33]: Is this comment really needed?

**Review by adinn — APPROVED** [review]
Looks good.

File: src/hotspot/share/services/nmtPreInit.hpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/share/services/nmtPreInit.hpp:33]: > Is this comment really needed?

Probably not. I will remove the comment.

**Comment by tstuefe** [comment]
> Looks good.

Thanks Andrew!

File: test/hotspot/gtest/nmt/test_nmtpreinit.cpp [hunk 2]
```diff
@@ -112,7 +113,7 @@ class TestAllocations {
     log_state();
   }
   void test_post() {
-    assert(NMTPreInit::in_preinit_phase() == false,
+    assert(!MemTracker::is_initialized() == false,
```
Comment (iklam) [inline test/hotspot/gtest/nmt/test_nmtpreinit.cpp:116]: Hmm, I am confused by the double negation :-)

File: src/hotspot/share/services/nmtPreInit.cpp [hunk 3]
```diff
@@ -180,8 +179,7 @@ void* NMTPreInit::do_os_malloc(size_t size) {
 // Switches from NMT pre-init state to NMT post-init state;
 //  in post-init, no modifications to the lookup table are possible.
 void NMTPreInit::pre_to_post() {
-  assert(_nmt_was_initialized == false, "just once");
-  _nmt_was_initialized = true;
+  assert(MemTracker::is_initialized() == false, "just once");
```
Comment (iklam) [inline src/hotspot/share/services/nmtPreInit.cpp:182]: Maybe this should be `!MemTracker::is_initialized()` to be consistent with the rest of the code?

File: test/hotspot/gtest/nmt/test_nmtpreinit.cpp (hunk 2 above)
Comment (tstuefe) [reply test/hotspot/gtest/nmt/test_nmtpreinit.cpp:116]: :-) Right, I'll fix it

File: src/hotspot/share/services/nmtPreInit.cpp (hunk 3 above)
Comment (tstuefe) [reply src/hotspot/share/services/nmtPreInit.cpp]: Done.

**Comment by tstuefe** [comment]
@iklam are you okay with the last version?

**Comment by tstuefe** [comment]
Pinging @iklam ... anything more I should do? If not, could you approve please? Thanks!

(Patch ran through in our nightlies for about a week now, all green)

**Review by iklam — APPROVED** [review]
All good. Sorry for the delay.

**Comment by tstuefe** [comment]
Thanks @iklam! No worries, we all scramble for 20.

/integrate

**Pull request closed without merging**